gpflow/models/cglb.py

```diff
@@ -0,0 +1,359 @@
+        c = trisolve(LB, Ares, lower=True) / sigma
+        tmp1 = trisolve(L, Kus, lower=True)
+        tmp2 = trisolve(LB, tmp1, lower=True)
+        sgpr_mean = matmul(tmp2, c, transpose_a=True)
+
+        if full_cov:
+            var = (
+                kernel(xnew)
+                + matmul(tmp2, tmp2, transpose_a=True)
+                - matmul(tmp1, tmp1, transpose_a=True)
+            )
+            var = tf.tile(var[None, ...], [self.num_latent_gps, 1, 1])  # [P, N, N]
+        else:
+            var = (
+                kernel(xnew, full_cov=False)
+                + tf.reduce_sum(tf.square(tmp2), 0)
+                - tf.reduce_sum(tf.square(tmp1), 0)
+            )
+            var = tf.tile(var[:, None], [1, self.num_latent_gps])
+
+        mean = sgpr_mean + cg_mean + self.mean_function(xnew)
+        return mean, var
+
+    def predict_y(
+        self,
+        xnew: InputData,
+        full_cov: bool = False,
+        full_output_cov: bool = False,
+        cg_tolerance: Union[float, None] = 1e-3,
+    ) -> MeanAndVariance:
+        """
+        Compute the mean and variance of the held-out data at the input points.
+        """
+        if full_cov or full_output_cov:
+            # See https://github.com/GPflow/GPflow/issues/1461
+            raise NotImplementedError(
+                "The predict_y method currently supports only the argument values full_cov=False and full_output_cov=False"
+            )
+
+        f_mean, f_var = self.predict_f(
+            xnew, full_cov=full_cov, full_output_cov=full_output_cov, cg_tolerance=cg_tolerance
+        )
+        return self.likelihood.predict_mean_and_var(f_mean, f_var)
+
+    def predict_log_density(
+        self,
+        data: RegressionData,
+        full_cov: bool = False,
+        full_output_cov: bool = False,
+        cg_tolerance: Union[float, None] = 1e-3,
+    ) -> tf.Tensor:
+        """
+        Compute the log density of the data at the new data points.
+        """
+        if full_cov or full_output_cov:
+            raise NotImplementedError(
+                "The predict_log_density method currently supports only the argument values full_cov=False and full_output_cov=False"
+            )
+
+        x, y = data
+        f_mean, f_var = self.predict_f(
+            x, full_cov=full_cov, full_output_cov=full_output_cov, cg_tolerance=cg_tolerance
+        )
+        return self.likelihood.predict_log_density(f_mean, f_var, y)
+
+
+class NystromPreconditioner:
+    """
+    Preconditioner of the form :math:`Q=(Q_ff + σ²I)⁻¹`,
+    where L is lower triangular with :math: `LLᵀ = Kᵤᵤ` :math:`A = σ⁻²L⁻¹Kᵤₓ` and :math:`B = AAᵀ + I = LᵦLᵦᵀ`
+    """
+
+    def __init__(self, A, LB, sigma_sq):
+        self.A = A
+        self.LB = LB
+        self.sigma_sq = sigma_sq
+
+    def __call__(self, v):
+        """
+        Computes :math:`vQ^{-1}`. Note that this is implemented as multipication of a row vector on the right.
```

Comment:
Looks like this method returns two thing, but the docstring says only one thing.